Hazel/LocklessDtls/LocklessDtlsConnectionListener.cs

```diff
@@ -0,0 +1,453 @@
+                    }
+                    catch (SocketException sx)
+                    {
+                        message.Recycle();
+                        if (sx.SocketErrorCode == SocketError.NotConnected)
+                        {
+                            this.InvokeInternalError(HazelInternalErrors.ConnectionDisconnected);
+                            return;
+                        }
+
+                        this.Logger.WriteError("Socket Ex in ReceiveLoop: " + sx.Message);
+                        continue;
+                    }
+                    catch (Exception ex)
+                    {
+                        message.Recycle();
+                        this.Logger.WriteError("Stopped due to: " + ex.Message);
+                        return;
+                    }
+
+                    LocklessDtlsServerConnection connection;
+                    if (!this.allConnections.TryGetValue(remoteEP, out connection))
+                    {
+                        lock (this.allConnections)
+                        {
+                            if (!this.allConnections.TryGetValue(remoteEP, out connection))
+                            {
+                                connection = new LocklessDtlsServerConnection(this, (IPEndPoint)remoteEP, this.IPMode, this.Logger);
+                                if (!this.allConnections.TryAdd(remoteEP, connection))
+                                {
+                                    this.Logger.WriteError("Failed to add unique connection! This should never happen!");
+                                }
+                            }
+                        }
+                    }
+
+                    EnqueueMessageReceived(message, connection);
+                }
+            }
+        }
+
+        internal void EnqueueMessageReceived(MessageReader message, LocklessDtlsServerConnection connection)
+        {
+            connection.PacketsReceived.Enqueue(message);
+            this.receiveQueue.TryAdd(connection);
+        }
+
+        private void ProcessingLoop()
+        {
+            var myTid = Thread.CurrentThread.ManagedThreadId;
+            while (this.receiveQueue.TryTake(out var sender))
+            {
+                if (Interlocked.CompareExchange(ref sender.LockedBy, myTid, 0) != 0)
+                {
+                    continue;
+                }
+
+                while (sender.PacketsReceived.TryDequeue(out var message))
+                {
+                    try
+                    {
+                        this.ReadCallback(message, sender);
+                    }
+                    catch (Exception ex)
+                    {
+                        this.Logger.WriteError("Unhandled exception in ReadCallback: " + ex);
+                    }
+                }
+
+                while (sender.PacketsSent.TryDequeue(out var message))
```

Comment:
Just reiterating that I think these two receive and send loops will need maximum processing counts.

Also, while I think the monitors will take care of sleeping the thread occasionally, if the TryTake doesn't induce a Wait, a thread could spin quite a bit - wondering if it'll be necessarily to sleep the thread every 100 or 1000 total messages.